py_ecc/optimized_bls12_381/optimized_pairing.py

```diff
@@ -161,5 +161,15 @@ def pairing(Q: Optimized_Point3D[FQ2],
     return miller_loop(twist(Q), cast_point_to_fq12(P), final_exponentiate=final_exponentiate)
 
 
-def final_exponentiate(p: Optimized_Field) -> Optimized_Field:
-    return p ** ((field_modulus ** 12 - 1) // curve_order)
+exptable = [FQ12([0] * i + [1] + [0] * (11 - i)) ** field_modulus for i in range(12)]
+
+
+def exp_by_p(x: FQ12) -> FQ12:
+    return sum((table_entry * coeff for table_entry, coeff in zip(exptable, x.coeffs)), FQ12.zero())
+
+
+def final_exponentiate(p: FQ12) -> FQ12:
+    cofactor = (field_modulus ** 4 - field_modulus ** 2 + 1) // curve_order
```

Comment:
This line appears to not be tied to the function arguments, and thus maybe suitable to be lifted into a constant to avoid runtime computation cost.

Reply:
The runtime computation cost is negligible; do what makes sense from a readability perspective? Cofactor is not used anywhere else and having it there makes it easier for someone else reading the function to understand what it's doing, which would be my rationale for just leaving it there, but I'm fine either way.

Reply:
I'm fine leaving it as is.  Less lazy approach would have been to test it out myself.... but I didn't do that.  Just wanted to make sure we weren't leaving easy performance gains on the table.